# sourcing or executing this script in an open R session will generate the normalized gene expression
# matrices used for batch correction, along with the appropriate meta data for each study.

library(scran)
library(biomaRt)
library(scater)
library(SingleCellExperiment)

library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)    

##############
## GSE81076 ##
##############

# Download file from GEO.
gse81076 <- bfcrpath(bfc, "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81076/suppl/GSE81076%5FD2%5F3%5F7%5F10%5F17%2Etxt%2Egz")
gse81076.df <- readSparseCounts(gse81076)

# Construct the meta data from the cell names.
plate.id <- sub("_.*", "", colnames(gse81076.df))
donor.id <- sub("(D[0-9]{1,2}).*", "\\1", plate.id)
gse81076.meta <- data.frame(Donor = donor.id, Plate=plate.id, Protocol="CELseq", Study="GSE81076")

# Remove superfluous suffixes from gene IDs.
gene.symbols <- sub("__chr([0-9XYM]+)", "", rownames(gse81076.df))
new.names <- uniquifyFeatureNames(seq_len(nrow(gse81076.df)), gene.symbols)
rownames(gse81076.df) <- new.names

# Create a SingleCellExperiment object.
sce.gse81076 <- SingleCellExperiment(list(counts=gse81076.df), 
    colData=gse81076.meta, rowData=DataFrame(Symbol=gene.symbols))
isSpike(sce.gse81076, "ERCC") <- grep("^ERCC-[0-9]*$", rownames(sce.gse81076))

# Remove low-quality cells.
sce.gse81076 <- calculateQCMetrics(sce.gse81076, compact=TRUE)

discard <- isOutlier(sce.gse81076$scater_qc$all$total_features_by_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse81076$scater_qc$all$total_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse81076$scater_qc$feature_control_ERCC$pct_counts, type="higher", nmads=3)
##summary(discard)
sce.gse81076 <- sce.gse81076[,!discard]

# Compute normalization factors.
set.seed(1000)
clusters <- quickCluster(sce.gse81076, min.mean=0.1, method="igraph")
##table(clusters)
sce.gse81076 <- computeSumFactors(sce.gse81076, clusters=clusters, min.mean=0.1)
##summary(sizeFactors(sce.gse81076))

sce.gse81076 <- computeSpikeFactors(sce.gse81076, general.use=FALSE)
sce.gse81076 <- normalize(sce.gse81076)
saveRDS(file="sce.gse81076.rds", sce.gse81076)

# Detect highly variable genes.
fit <- trendVar(sce.gse81076, block=sce.gse81076$Plate)
dec <- decomposeVar(sce.gse81076, fit)
##plot(dec$mean, dec$var)
##points(fit$means, fit$vars, col="red")
##curve(fit$trend(x), col="dodgerblue")
saveRDS(file="dec.gse81076.rds", dec)

# Clear environment and invoke garbage collector
rm(list=ls())
gc()

##############
## GSE85241 ##
##############

# Download file from GEO.
gse85241 <- bfcrpath(bfc, "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE85nnn/GSE85241/suppl/GSE85241%5Fcellsystems%5Fdataset%5F4donors%5Fupdated%2Ecsv%2Egz")
gse85241.df <- readSparseCounts(gse85241, quote='"')

# Construct the meta data from the cell names.
plate.id <- sub("_.*", "", colnames(gse85241.df))
donor.id <- sub("-.*", "", plate.id)
gse85241.meta <- data.frame(Donor = donor.id, Plate=plate.id, Protocol="CELseq2", Study="GSE85241")

# Remove superfluous suffixes from gene IDs.
gene.symbols <- sub("__chr([0-9XYM]+)", "", rownames(gse85241.df))
new.names <- uniquifyFeatureNames(seq_len(nrow(gse85241.df)), gene.symbols)
rownames(gse85241.df) <- new.names

# Create a SingleCellExperiment object.
sce.gse85241 <- SingleCellExperiment(list(counts=gse85241.df), 
    colData=gse85241.meta, rowData=DataFrame(Symbol=gene.symbols))
isSpike(sce.gse85241, "ERCC") <- grep("^ERCC-[0-9]*$", rownames(sce.gse85241))

# Remove low-quality cells.
sce.gse85241 <- calculateQCMetrics(sce.gse85241, compact=TRUE)

discard <- isOutlier(sce.gse85241$scater_qc$all$total_features_by_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse85241$scater_qc$all$total_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse85241$scater_qc$feature_control_ERCC$pct_counts, type="higher", nmads=3)
##summary(discard)
sce.gse85241 <- sce.gse85241[,!discard]

# Compute normalization factors.
set.seed(1000)
clusters <- quickCluster(sce.gse85241, min.mean=0.1, method="igraph")
##table(clusters)
sce.gse85241 <- computeSumFactors(sce.gse85241, clusters=clusters, min.mean=0.1)
##summary(sizeFactors(sce.gse85241))

sce.gse85241 <- computeSpikeFactors(sce.gse85241, general.use=FALSE)
sce.gse85241 <- normalize(sce.gse85241)
saveRDS(file="sce.gse85241.rds", sce.gse85241)

# Detect highly variable genes.
fit <- trendVar(sce.gse85241, block=sce.gse85241$Plate)
dec <- decomposeVar(sce.gse85241, fit)
##plot(dec$mean, dec$var)
##points(fit$means, fit$vars, col="red")
##curve(fit$trend(x), col="dodgerblue")
saveRDS(file="dec.gse85241.rds", dec)

# Clear environment and invoke garbage collector
rm(list=ls())
gc()

##############
## GSE86473 ##
##############

# The raw/processed counts table was not available for download from GEO for this data set.
# Instead, data were generated by mapping the original fastq's back to mm10, 
# then using featureCounts to quantify against mm10 ensembl build 86.
host.path <- "https://jmlab-gitlab.cri.camres.org/aaron/MNNDataFiles/raw/2d5ee3484990feb18e051bb7323bafd42b095b84"
alpha <- bfcrpath(bfc, file.path(host.path, 'alpha-feature_counts.tsv.gz'))
beta <- bfcrpath(bfc, file.path(host.path, 'beta-feature_counts.tsv.gz'))
delta <- bfcrpath(bfc, file.path(host.path, 'delta-feature_counts.tsv.gz'))
pp <- bfcrpath(bfc, file.path(host.path, 'PP-feature_counts.tsv.gz'))

alpha.df <- readSparseCounts(alpha)
beta.df <- readSparseCounts(beta)
delta.df <- readSparseCounts(delta)
pp.df <- readSparseCounts(pp)

stopifnot(identical(rownames(alpha.df), rownames(beta.df)))
stopifnot(identical(rownames(alpha.df), rownames(delta.df)))
stopifnot(identical(rownames(alpha.df), rownames(pp.df)))

# Clean up row and column names. 
gse86473.df <- cbind(alpha.df, beta.df, delta.df, pp.df)
rownames(gse86473.df) <- sub("ERCC\\.", "ERCC-", rownames(gse86473.df))

samp.names <- sub("\\.dedup$", "", colnames(gse86473.df))
colnames(gse86473.df) <- tolower(samp.names)

# Load in the metadata.
meta.fname <- bfcrpath(bfc, file.path(host.path, 'GSE86473_experimental_design.tsv'))
gse86473.meta <- read.delim(meta.fname, stringsAsFactors=FALSE)
gse86473.meta$Study <- "GSE86473"
gse86473.meta$CellType <- paste(toupper(substr(gse86473.meta$CellType, 1, 1)),
                                substring(gse86473.meta$CellType, 2), sep="")

# Pulling out gene symbols.
library(EnsDb.Hsapiens.v86)
symb <- mapIds(EnsDb.Hsapiens.v86, keys=rownames(gse86473.df), keytype="GENEID", column="SYMBOL")

# Creating a SingleCellExperiment object.
sce.gse86473 <- SingleCellExperiment(list(counts=gse86473.df), 
    colData=gse86473.meta, rowData=DataFrame(Symbol=symb))
isSpike(sce.gse86473, "ERCC") <- grep("^ERCC-", rownames(sce.gse86473))

# Remove low-quality cells.
sce.gse86473 <- calculateQCMetrics(sce.gse86473, compact=TRUE)

discard <- isOutlier(sce.gse86473$scater_qc$all$total_features_by_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse86473$scater_qc$all$total_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.gse86473$scater_qc$feature_control_ERCC$pct_counts, type="higher", nmads=3)
##summary(discard)
sce.gse86473 <- sce.gse86473[,!discard]

# Compute normalization factors.
set.seed(1000)
clusters <- quickCluster(sce.gse86473, min.mean=1, method="igraph")
##table(clusters)
sce.gse86473 <- computeSumFactors(sce.gse86473, clusters=clusters, min.mean=1)
##summary(sizeFactors(sce.gse86473))

#sce.gse86473 <- computeSpikeFactors(sce.gse86473, general.use=FALSE) # As there are actually no spike-in counts, anywhere.
sce.gse86473 <- normalize(sce.gse86473)
saveRDS(file="sce.gse86473.rds", sce.gse86473)

# Detect highly variable genes (nothing obvious to block on, unfortunately).
fit <- trendVar(sce.gse86473, use.spikes=FALSE, loess.args=list(span=0.05))
##plot(fit$mean, fit$var)
##curve(fit$trend(x), col="dodgerblue")
dec <- decomposeVar(sce.gse86473, fit)
saveRDS(file="dec.gse86473.rds", dec)

# Clear environment and invoke garbage collector.
rm(list=ls())
gc()

#################
## E-MTAB-5061 ##
#################

# Unpacking the file:
# - First 2 columns are gene symbol and NCBI ID.
# - Next X columns are the RPKMs.
# - Remaining X columns are the counts.

emat <- bfcrpath(bfc, "https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-5061/files/E-MTAB-5061.processed.1.zip")
col.names <- read.table(unz(emat, "pancreas_refseq_rpkms_counts_3514sc.txt"), header=FALSE, sep="\t", stringsAsFactors=FALSE, comment.char="", nrows = 1)[,-1]
ncells <- length(col.names)

what <- vector("list", ncells*2 + 2)
what[[1]] <- "character"
what[seq_len(ncells) + ncells + 2] <- "integer"

emtab.df <- read.table(unz(emat, "pancreas_refseq_rpkms_counts_3514sc.txt"), header=FALSE, sep="\t", stringsAsFactors=FALSE, colClasses=what, skip=1)
emtab.df <- emtab.df[!duplicated(emtab.df[,1]),]
row.names <- emtab.df[,1]
emtab.df <- emtab.df[,-1]

rownames(emtab.df) <- row.names
colnames(emtab.df) <- col.names

# Reading in the metadata and constructing the appropriate meta data columns, i.e. donor, plate, protocol, study
meta.fname <- bfcrpath(bfc, "https://www.ebi.ac.uk/arrayexpress/files/E-MTAB-5061/E-MTAB-5061.sdrf.txt")
emtab.sdrf <- read.delim(meta.fname, stringsAsFactors=FALSE)
emtab.meta <- emtab.sdrf[, c("Assay.Name", "Characteristics.cell.type.", "Characteristics.individual.")]
colnames(emtab.meta) <- c("Sample", "CellType", "Donor")
emtab.meta$Study <- "E-MTAB-5061"
emtab.meta$Protocol <- "SmartSeq2"

# Editing the cell type labels.
emtab.meta$CellType <- gsub(" cell", "", emtab.meta$CellType)
emtab.meta$CellType <- paste(toupper(substr(emtab.meta$CellType, 1, 1)),
                             substring(emtab.meta$CellType, 2), sep="")

# Creating a SingleCellExperiment object here.
sce.emtab <- SingleCellExperiment(list(counts=as.matrix(emtab.df)), 
    colData=emtab.meta[match(colnames(emtab.df), emtab.meta$Sample),])
isSpike(sce.emtab, "ERCC") <- grep("^ERCC_", rownames(sce.emtab))

# Remove the marked low quality cells
low.qual <- emtab.sdrf$Assay.Name[emtab.sdrf$Characteristics.single.cell.well.quality. == "low quality cell"]
sce.emtab <- sce.emtab[,!colnames(sce.emtab) %in% low.qual] 

# Remove more low quality cells.
sce.emtab <- calculateQCMetrics(sce.emtab, compact=TRUE)

discard <- isOutlier(sce.emtab$scater_qc$all$total_features_by_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.emtab$scater_qc$all$total_counts, log=TRUE, type="lower", nmads=3) |
    isOutlier(sce.emtab$scater_qc$feature_control_ERCC$pct_counts, type="higher", nmads=3)
##summary(discard)
sce.emtab <- sce.emtab[,!discard]

# Compute normalization factors.
set.seed(1000)
clusters <- quickCluster(sce.emtab, min.mean=1, method="igraph")
##table(clusters)
sce.emtab <- computeSumFactors(sce.emtab, clusters=clusters, min.mean=1)
##summary(sizeFactors(sce.emtab))

sce.emtab <- computeSpikeFactors(sce.emtab, general.use=FALSE) 
sce.emtab <- normalize(sce.emtab)
saveRDS(file="sce.emtab5601.rds", sce.emtab)

# Detect highly variable genes.
for.hvg <- sce.emtab[,sizeFactors(sce.emtab, "ERCC") > 0 & sce.emtab$Donor!="AZ"]
for.hvg <- multiBlockNorm(for.hvg, for.hvg$Donor) 
comb.out <- multiBlockVar(for.hvg, for.hvg$Donor)
##all.donors <- unique(for.hvg$Donor)
##par(mfrow=c(ceiling(length(all.donors)/3),3))
##is.spike <- isSpike(for.hvg)
##for (plate in all.donors) {
##    cur.out <- comb.out$per.block[[plate]]
##    plot(cur.out$mean, cur.out$total, pch=16, cex=0.6, xlab="Mean log-expression", 
##         ylab="Variance of log-expression", main=plate)
##    curve(metadata(cur.out)$trend(x), col="dodgerblue", lwd=2, add=TRUE)
##    points(cur.out$mean[is.spike], cur.out$total[is.spike], col="red", pch=16)
##}
saveRDS(file="dec.emtab5601.rds", comb.out)

# Clear environment and invoke garbage collector.
rm(list=ls())
gc()
